pkg/macaroon/macaroon.go

```diff
@@ -0,0 +1,133 @@
+// Copyright (C) 2019 Storj Labs, Inc. // See LICENSE for copying information.
+
+package macaroon
+
+import (
+	"crypto/hmac"
+	"crypto/rand"
+	"crypto/sha256"
+	"crypto/subtle"
+)
+
+// Macaroon is a struct that determine contextual caveats and authorization
+type Macaroon struct {
+	head    []byte
+	caveats []Caveat
+	tail    []byte
+}
+
+// Caveat is a struct that determine restriction
+type Caveat struct {
+	Identifier string
+}
+
+// NewUnrestricted creates Macaroon with random Head and generated Tail
+func NewUnrestricted(identifier []byte, secret []byte) *Macaroon {
+	return &Macaroon{
+		head: identifier,
+		tail: sign(secret, identifier),
+	}
+}
+
+// sign
+func sign(secret []byte, data []byte) []byte {
+	signer := hmac.New(sha256.New, secret)
+	// Error skipped because sha256 does not return error
+	_, _ = signer.Write(data)
+
+	return signer.Sum(nil)
+}
+
+// NewSecret generates cryptographically random 32 bytes
+func NewSecret() (secret []byte, err error) {
+	secret = make([]byte, 32)
+
+	_, err = rand.Read(secret)
+	if err != nil {
+		return nil, err
+	}
+
+	return secret, nil
+}
+
+// NewNonce generates cryptographically random 32 bytes
+func NewNonce() (nonce []byte, err error) {
```

Comment:
since NewSecret() and NewNonce() are identical, maybe just call both NewRand or something? or call both NewSecret()?

Reply:
agree, we can just use another secret as nonce